source/common/conn_pool/conn_pool_base.cc

```diff
@@ -608,11 +660,27 @@ void ConnPoolImplBase::onPendingStreamCancel(PendingStream& stream,
     state_.decrPendingStreams(1);
     stream.removeFromList(pending_streams_);
   }
-  if (policy == Envoy::ConnectionPool::CancelPolicy::CloseExcess && !connecting_clients_.empty() &&
-      connectingConnectionIsExcess()) {
-    auto& client = *connecting_clients_.front();
-    transitionActiveClientState(client, ActiveClient::State::Draining);
-    client.close();
+  if (policy == Envoy::ConnectionPool::CancelPolicy::CloseExcess) {
+    if (!connecting_clients_.empty() &&
+        connectingConnectionIsExcess(*connecting_clients_.front())) {
+      auto& client = *connecting_clients_.front();
+      transitionActiveClientState(client, ActiveClient::State::Draining);
+      client.close();
+    } else if (!early_data_clients_.empty()) {
+      ActiveClient* idle_client{nullptr};
+      for (ActiveClientPtr& client : early_data_clients_) {
+        if (client->numActiveStreams() == 0) {
+          if (connectingConnectionIsExcess(*client)) {
+            idle_client = client.get();
+          }
+          break;
```

Comment:
Does this mean that under some circumstance we won't actually move a client to draining when this method is called?

Reply:
When a pending stream get cancelled, we only drain close an idle client if without it the pool can still serve the rest pending streams once all the clients are connected. This logic is in connectingConnectionIsExcess().